cmd/signing-server/main.go

```diff
@@ -69,14 +73,126 @@ type Config struct {
 	DisableAuth        bool
 	DisableHTTPS       bool
 
+	HSMModule   string
+	HSMSlot     int
+	HSMPass     string
+	HSMKeyLabel string
+	HSMKeyId    string
+
+	HSMContext   *pkcs11.Ctx
+	HSMSession   pkcs11.SessionHandle
+	HSMKeyHandle pkcs11.ObjectHandle
+
 	// calculated by program
 	Logger *zap.Logger
 }
 
+func (c *Config) SetupHSM() error {
+	c.Logger.Info("setup HSM signing", zap.String("module", c.HSMModule))
+	p := pkcs11.New(c.HSMModule)
+	err := p.Initialize()
```

Comment:
in the spirit of defensive programming, we might want to check whether `p != nil` before calling `p.Initialize()`

```go
// New creates a new context and initializes the module/library for use.
func New(module string) *Ctx {
	c := new(Ctx)
	mod := C.CString(module)
	defer C.free(unsafe.Pointer(mod))
	c.ctx = C.New(mod)
	if c.ctx == nil {
		return nil
	}
	return c
}
```

Reply:
This uses cgo and is just memory allocation, thus I vote against a nil check

Reply:
I am not that familiar w/ golang, however if I correctly understand [this](https://github.com/miekg/pkcs11/blob/master/pkcs11.go#L47) is where the called `New` function is defined. If this is correct, then it looks to me as if there were codepaths that might return `NULL` (which I guess is `nil` in vanilla-golang?)

Which might be a reason in favour of nil-checking. WDYT?

Reply:
Just checked again. fair enough, if the LoadLibrary syscall from win32 fails the handle returned is NULL. since this is before initialize. a nil check is mandatory here. thanks for the pointer.

Reply:
done